File: lib/collect/argocd.ts
import { getInventoryConfig } from "@/lib/collect/config"
import type { CollectAdapterResult } from "@/lib/collect/adapters"
import type { ArgoCdApplication, CollectErrorCode } from "@/lib/collect/types"

type ArgoApplicationListResponse = {
  items?: ArgoApplicationItem[]
}

type ArgoApplicationItem = {
  metadata?: {
    name?: string
    namespace?: string
  }
  status?: {
    sync?: {
      status?: string
      revision?: string
    }
    health?: {
      status?: string
    }
  }
}

export async function collectArgoCd(
  signal: AbortSignal
): Promise<CollectAdapterResult<"argocd">> {
  const config = getInventoryConfig()
  const baseUrl = config.argocd?.baseUrl?.replace(/\/$/, "") ?? ""
  const collectedAt = new Date().toISOString()

  if (!baseUrl) {
    return {
      status: "unknown",
      collectedAt,
      stale: false,
      error: {
        code: "UNKNOWN_ERROR",
        message: "Argo CD base URL not configured",
      },
      data: { applications: [] },
    }
  }

  const token = process.env.ARGOCD_TOKEN?.trim()
  if (!token) {
    return {
      status: "permission_error",
      collectedAt,
      stale: false,
      error: {
        code: "PERMISSION_DENIED",
        message: "ARGOCD_TOKEN is missing",
      },
      data: { applications: [] },
    }
  }

  try {
    const response = await fetch(`${baseUrl}/api/v1/applications`, {
      headers: {
        Accept: "application/json",
        Authorization: `Bearer ${token}`,
      },
      signal,
    })

    if (!response.ok) {
      return {
        status: response.status === 401 || response.status === 403 ? "permission_error" : "down",
        collectedAt,
        stale: false,
        error: {
          code:
            response.status === 401 || response.status === 403
              ? "PERMISSION_DENIED"
              : "CONNECTION_FAILED",
          message: `Argo CD API responded with status ${response.status}`,
        },
        data: { applications: [] },
      }
    }

    const payload = (await response.json()) as ArgoApplicationListResponse
    const applications = (payload.items ?? []).flatMap((item) =>
      toApplication(item, baseUrl)
    )

    return {
      status: "ok",
      collectedAt,
      stale: false,
      error: null,
      data: { applications },
    }
  } catch (error) {
    const mappedError = mapError(error)
    return {
      status: mappedError.code === "TIMEOUT" ? "timeout" : "down",
      collectedAt,
      stale: false,
      error: mappedError,
      data: { applications: [] },
    }
  }
}

function toApplication(
  item: ArgoApplicationItem,
  baseUrl: string
): ArgoCdApplication[] {
  const name = item.metadata?.name
  const namespace = item.metadata?.namespace

  if (!name || !namespace) {
    return []
  }

  return [
    {
      name,
      namespace,
      syncStatus: toSyncStatus(item.status?.sync?.status),
      healthStatus: toHealthStatus(item.status?.health?.status),
      revision: item.status?.sync?.revision ?? null,
      link: `${baseUrl}/applications/${encodeURIComponent(name)}`,
    },
  ]
}

function toSyncStatus(value?: string): ArgoCdApplication["syncStatus"] {
  if (value === "Synced" || value === "OutOfSync") {
    return value
  }
  return "Unknown"
}

function toHealthStatus(value?: string): ArgoCdApplication["healthStatus"] {
  if (
    value === "Healthy" ||
    value === "Progressing" ||
    value === "Degraded"
  ) {
    return value
  }
  return "Unknown"
}

function mapError(error: unknown): { code: CollectErrorCode; message: string } {
  if (error instanceof Error && error.name === "AbortError") {
    return {
      code: "TIMEOUT",
      message: "Argo CD API request timed out",
    }
  }

  return {
    code: "CONNECTION_FAILED",
    message:
      error instanceof Error ? error.message : "Argo CD API request failed",
  }
}

File: backend/internal/collector/argocd.go
package collector

import (
	"context"
	"encoding/json"
	"net/http"
	"omni-backend/internal/models"
	"strings"
	"sync"
	"time"
)

type argoApplicationListResponse struct {
	Items []argoApplicationItem `json:"items"`
}

type argoApplicationItem struct {
	Metadata struct {
		Name      string `json:"name"`
		Namespace string `json:"namespace"`
	} `json:"metadata"`
	Status struct {
		Sync struct {
			Status   string `json:"status"`
			Revision string `json:"revision"`
		} `json:"sync"`
		Health struct {
			Status string `json:"status"`
		} `json:"health"`
	} `json:"status"`
}

func CollectArgoCD(ctx context.Context, targets []models.ArgoCDCollectTarget) models.CollectEnvelope[models.ArgoCdData] {
	now := time.Now().Format(time.RFC3339)

	if len(targets) == 0 {
		collectedAt := now
		return models.CollectEnvelope[models.ArgoCdData]{
			Source:      models.SourceArgoCD,
			Status:      models.StatusUnknown,
			AttemptedAt: now,
			CollectedAt: &collectedAt,
			Stale:       false,
			Data:        models.ArgoCdData{Applications: []models.ArgoCdApplication{}},
		}
	}
	var wg sync.WaitGroup
	var mu sync.Mutex
	applications := []models.ArgoCdApplication{}
	status := models.StatusOk
	var collectErr *models.CollectError

	for _, target := range targets {
		wg.Add(1)
		go func(target models.ArgoCDCollectTarget) {
			defer wg.Done()
			result := collectArgoCDTarget(ctx, target, now)
			mu.Lock()
			applications = append(applications, result.Data.Applications...)
			if severity(result.Status) > severity(status) {
				status = result.Status
				collectErr = result.Error
			}
			mu.Unlock()
		}(target)
	}
	wg.Wait()

	collectedAt := now
	return models.CollectEnvelope[models.ArgoCdData]{
		Source:      models.SourceArgoCD,
		Status:      status,
		AttemptedAt: now,
		CollectedAt: &collectedAt,
		Stale:       status == models.StatusStale,
		Error:       collectErr,
		Data:        models.ArgoCdData{Applications: applications},
	}
}

func collectArgoCDTarget(ctx context.Context, target models.ArgoCDCollectTarget, now string) models.CollectEnvelope[models.ArgoCdData] {
	baseUrl := strings.TrimRight(target.BaseURL, "/")

	reqCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, "GET", baseUrl+"/api/v1/applications", nil)
	if err != nil {
		return argoError(now, models.ErrConnectionFailed, err.Error(), models.StatusDown)
	}

	req.Header.Set("Accept", "application/json")
	req.Header.Set("Authorization", "Bearer "+target.Token)

	client := &http.Client{}
	resp, err := client.Do(req)

	if err != nil {
		code := models.ErrConnectionFailed
		status := models.StatusDown
		if reqCtx.Err() == context.DeadlineExceeded {
			code = models.ErrTimeout
			status = models.StatusTimeout
		}
		return argoError(now, code, err.Error(), status)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		code := models.ErrConnectionFailed
		status := models.StatusDown
		if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
			code = models.ErrPermissionDenied
			status = models.StatusPermissionError
		}
		upstreamStatus := resp.StatusCode
		return argoError(now, code, "Argo CD API responded with error", status, &upstreamStatus)
	}

	var payload argoApplicationListResponse
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return argoError(now, models.ErrUnknownError, "Failed to parse ArgoCD response", models.StatusDown)
	}

	var applications []models.ArgoCdApplication
	isStale := false
	isProgressing := false

	for _, item := range payload.Items {
		name := item.Metadata.Name
		namespace := item.Metadata.Namespace
		if name == "" || namespace == "" {
			continue
		}

		syncStatus := item.Status.Sync.Status
		if syncStatus != "Synced" && syncStatus != "OutOfSync" {
			syncStatus = "Unknown"
		}

		healthStatus := item.Status.Health.Status
		if healthStatus != "Healthy" && healthStatus != "Progressing" && healthStatus != "Degraded" {
			healthStatus = "Unknown"
		}

		rev := item.Status.Sync.Revision
		var revPtr *string
		if rev != "" {
			revPtr = &rev
		}

		applications = append(applications, models.ArgoCdApplication{
			IntegrationName: target.Name,
			Name:            name,
			Namespace:       namespace,
			SyncStatus:      syncStatus,
			HealthStatus:    healthStatus,
			Revision:        revPtr,
			Link:            baseUrl + "/applications/" + name,
		})

		if healthStatus == "Degraded" || healthStatus == "Unknown" || (syncStatus != "Synced" && healthStatus != "Progressing") {
			isStale = true
		}
		if healthStatus == "Progressing" {
			isProgressing = true
		}
	}

	status := models.StatusOk
	if isStale {
		status = models.StatusStale
	} else if isProgressing {
		status = models.StatusProgressing
	}

	collectedAt := now
	return models.CollectEnvelope[models.ArgoCdData]{
		Source:      models.SourceArgoCD,
		Status:      status,
		AttemptedAt: now,
		CollectedAt: &collectedAt,
		Stale:       false,
		Error:       nil,
		Data:        models.ArgoCdData{Applications: applications},
	}
}

func severity(status models.SourceStatus) int {
	switch status {
	case models.StatusDown, models.StatusTimeout, models.StatusPermissionError:
		return 4
	case models.StatusStale:
		return 3
	case models.StatusProgressing:
		return 2
	case models.StatusUnknown:
		return 1
	default:
		return 0
	}
}

func argoError(now string, code models.CollectErrorCode, msg string, status models.SourceStatus, upstreamStatus ...*int) models.CollectEnvelope[models.ArgoCdData] {
	var upstream *int
	if len(upstreamStatus) > 0 {
		upstream = upstreamStatus[0]
	}
	return models.CollectEnvelope[models.ArgoCdData]{
		Source:      models.SourceArgoCD,
		Status:      status,
		AttemptedAt: now,
		Stale:       false,
		Error:       &models.CollectError{Code: code, Message: msg, UpstreamStatus: upstream},
		Data:        models.ArgoCdData{Applications: []models.ArgoCdApplication{}},
	}
}
